### PR TITLE
Add combinatorial data attribute

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Adds the CombinatorialDataAttribute for use in S6932

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.TestFramework/Common/CombinatorialDataAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+public sealed class CombinatorialDataAttribute : Attribute, ITestDataSource
+{
+    public IEnumerable<object[]> GetData(MethodInfo methodInfo)
+    {
+        var valuesPerParameter = methodInfo.GetParameters().Select(x => x.GetCustomAttribute<DataValuesAttribute>()?.Values
+            ?? throw new InvalidOperationException("Combinatorial test requires all parameters to have the [DataValues] attribute set")).ToArray();
+        var parameterIndices = new int[valuesPerParameter.Length];
+
+        while (true)
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Common/CombinatorialDataAttribute.cs:46]: I don't like that quasi-infinite loop. How about something like this?

```cs
public IEnumerable<object[]> GetData(MethodInfo methodInfo)
{
    var valuesPerParameter = methodInfo.GetParameters().Select(p => p.GetCustomAttribute<DataValuesAttribute>()?.Values
        ?? throw new InvalidOperationException("Combinatorial test requires all parameters to have the [DataValues] attribute set")).ToArray();
    var arrayLengths = valuesPerParameter.Select(x => x.Length).ToArray();
    var totalCombinations = arrayLengths.Aggregate(1, (x, y) => x * y);
    var parameterIndices = new int[arrayLengths.Length];

    for (var i = 0; i < totalCombinations; i++)
    {
        var temp = i;
        for (var j = arrayLengths.Length - 1; j >= 0; j--)
        {
            parameterIndices[j] = temp % arrayLengths[j];
            temp /= arrayLengths[j];
        }
        yield return CreateArguments(valuesPerParameter, parameterIndices);
    }
}
```

**Review by zsolt-kolbay-sonarsource — CHANGES_REQUESTED** [review]
Looks good. Left some minor comments. Rebase is needed

File: analyzers/tests/SonarAnalyzer.TestFramework/Common/CombinatorialDataAttribute.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.TestFramework/Common/CombinatorialDataAttribute.cs]: This is so much better. Thanks for taking the time to come up with this. I tried and failed to do so.

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**